optaplanner-core/src/main/java/org/optaplanner/core/impl/util/Sequence.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.util;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class Sequence<T> {
+    private final TreeSet<T> consecutiveItemsSet;
+    private final Map<T, Integer> count;
+    private final ConsecutiveSetTree<T, ?, ?> sourceTree;
+
+    protected Sequence(ConsecutiveSetTree<T, ?, ?> sourceTree) {
+        this(sourceTree, new TreeSet<>(sourceTree.comparator), new IdentityHashMap<>());
+    }
+
+    protected Sequence(ConsecutiveSetTree<T, ?, ?> sourceTree, TreeSet<T> consecutiveItemsSet, Map<T, Integer> count) {
+        this.sourceTree = sourceTree;
+        this.consecutiveItemsSet = consecutiveItemsSet;
+        this.count = count;
+    }
+
+    public TreeSet<T> getItems() {
+        return consecutiveItemsSet;
+    }
+
+    public int getLength() {
+        return consecutiveItemsSet.size();
+    }
+
+    protected boolean isEmpty() {
+        return consecutiveItemsSet.isEmpty();
+    }
+
+    protected int getCountIncludingDuplicates() {
+        return count.values().stream().reduce(Integer::sum).orElse(0);
+    }
+
+    protected Stream<T> getDuplicatedStream() {
+        return consecutiveItemsSet.stream()
+                .flatMap(item -> IntStream.range(0, count.get(item)).mapToObj(index -> item));
+    }
```

Comment:
I am a bit blurry as to what we need this for. But I really do not like the fact that it is not incremental.
My view of `Sequence` is that is just a dumb no-op view to the underlying collection. (See the idea of a timeline I tried to explain earlier.)